### ISSUE TITLE
LineEdit and TextEdit to optionally not capture KEY_TAB (=ui_focus_next)

### ISSUE THREAD

**Issue by goatchurchprime** [godotengine/godot-proposals#3050]
### Describe the project you are working on

VR cave survey with a keyboard driven UI interface

### Describe the problem or limitation you are having in your project

If you set up a 2D Viewport with a user interface made from buttons and boxes, you can quickly TAB and SHIFT_TAB around it in the keyboard to reach the different options without touching the mouse.

This works until one of the nodes is a TextEdit or LineEdit, and then your TAB key will simply insert tab characters into the text and you can never escape from this part of the UI without resorting to clicking on something else with the mouse.

I note that parts of the Godot Editor do not suffer from this (I can tab across the different boxes on the left such as the "Filter nodes" or "Search files" TextLines or on the left hand pane of the editor), so this feature is used.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Given that it is present in the editor, a programmer expects the feature to be available, so if you're like me you waste quite a bit of time looking for it.  If it's not supposed to be here, it should say so in the documentation so we can get straight to the ugly work-around (see below).  

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

You could probably implement it as a FocusMode: FOCUS_DO_NOT_CAPTURE_TAB_KEYS option

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This has one work-around, where you have to attach a script to each text box that captures and implements the TAB key, but it's pretty ugly, especially if you have many editor boxes.
https://www.reddit.com/r/godot/comments/fvtwey/how_to_avoid_textedit_capturing_tab/

```GDScript
func _input(event):
	if event.is_action_pressed("ui_focus_prev") and has_focus():
		if focus_previous != "":
			get_node(focus_previous).grab_focus()
		get_tree().set_input_as_handled()
	elif event.is_action_pressed("ui_focus_next") and has_focus():
		if focus_next != "":
			get_node(focus_next).grab_focus()
		get_tree().set_input_as_handled()
```

### Is there a reason why this should be core and not an add-on in the asset library?

This is a small feature on a core node.

**Comment by Calinou** [comment]
You should be able to press Ctrl + Tab to focus to the next control while focused on a LineEdit or TextEdit node. That's how it works in the script editor at least.

**Comment by manuel3108** [comment]
Here is the same implementation, but in C#

```c#
        if (@event.IsActionPressed("ui_focus_next") && HasFocus())
        {
            if (!string.IsNullOrEmpty(FocusNext))
            {
                ((Control)GetNode(FocusNext)).GrabFocus();
                GetTree().Root.SetInputAsHandled();
            }
        }
        else if (@event.IsActionPressed("ui_focus_next") && HasFocus())
        {
            if (!string.IsNullOrEmpty(FocusPrevious))
            {
                ((Control)GetNode(FocusPrevious)).GrabFocus();
                GetTree().Root.SetInputAsHandled();
            }
        }
```

**Comment by Giganzo** [comment]
> This is resolved for LineEdit - when you press Escape while editing, you will exit editing mode, but the LineEdit will still be focused. This allows to use Tab or arrows to change focus.
> 
> I think TextEdit should also get this functionality. It will work nice at least in the inspector.

Yeah I thought about that too, and it could be nice. Maybe have some time to look at that later.
However there is i a potential problem (currently happens for LineEdit). Say you open a dialog like the big text for inspector TextEdit's or TextEdit's in Project Settings. Now Escape will close the dialog.

**Comment by KoBeWi** [comment]
That can be resolved. The LineEdit/TextEdit needs to properly handle event, so it's not propagated to the window.

**Comment by Giganzo** [comment]
> That can be resolved. The LineEdit/TextEdit needs to properly handle event, so it's not propagated to the window.

But would that not be confusing for the user when escape dose not do what they might expect?
And is this escape key behavior for LineEdit/TextEdit common anywhere else?

**Comment by KoBeWi** [comment]
> But would that not be confusing for the user when escape dose not do what they might expect?

It's not like they'd block the Escape forever. You just have to press it twice. If the user notices disappearing caret on first press then it's within expected behavior.

> And is this escape key behavior for LineEdit/TextEdit common anywhere else?

You mean outside Godot or in other Controls? In Godot we don't really have others with this behavior, because it's not needed. In other apps, I don't know. Browsers don't let you write tabs at all and it's annoying.

**Comment by kitbdev** [comment]
- TextEdit support added by https://github.com/godotengine/godot/pull/76829

TextEdit now has adds `tab_input_mode` to toggle tab input behavior, using the `ui_focus_mode` action `Ctrl+M`.

Also, it was always possible to unbind or change the built-in shortcut `ui_text_indent`, to prevent all TextEdits from taking tab events.

Though, `tab_input_mode` is not currently working for CodeEdit.

**Comment by Calinou** [comment]
> TextEdit now has adds `tab_input_mode` to toggle tab input behavior, using the `ui_focus_mode` action `Ctrl+M`.

Any opinions on setting `tab_input_mode` to `false` for TextEdit nodes newly created in the editor (using an approach similar to https://github.com/godotengine/godot-proposals/issues/14750 for compatibility)? This would better match the behavior in other applications for multiline text fields not intended for code. (CodeEdit would still have tab-to-indent enabled by default.)

**Comment by kitbdev** [comment]
Ideally TextEdit would have a separate editing mode instead like LineEdit has, so the interaction is the same in places like the Inspector. (also mentioned https://github.com/godotengine/godot/pull/102663#issuecomment-2883917448)
